File: 프로그래머스/Lv.2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#include <malloc.h>

bool is_prime(long long n) {
    if (n <= 1) return false;
    else if (n == 2) return true;

    for (long long i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) return false;
    }

    return true;
}

char* to_knum(int n, int k) {
    int length = 0;
    int temp = n;

    while (temp > 0) {
        temp /= k;
        length++;
    }

    char* result = (char*)malloc((length + 1) * sizeof(char));

    int index = 0;
    while (n > 0) {
        result[length - 1 - index++] = n % k + '0';
        n /= k;
    }

    result[index] = '\0';

    return result;
}

char* substring(char* str, int start, int end) {
    int length = end - start;
    char* result = (char*)malloc(length + 1); 
    
    for (int i = 0; i < length; i++) {
        result[i] = str[start + i];
    }

    result[length] = '\0';

    return result;
}

int solution(int n, int k) {
    int answer = 0;
    char* s = to_knum(n, k);
    
    for (int i = 0, j; s[i] != '\0'; i = j) {
        for (j = i + 1; s[j] != '\0' && s[j] != '0'; j++);

        char* substring_result = substring(s, i, j);

        char* endptr;
        long long number = strtoll(substring_result, &endptr, 10);

        free(substring_result);

        if (is_prime(number)) {
            answer++;
        }
    }

    free(s);

    return answer;
}

File: 프로그래머스/Lv.2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.py
def is_prime(n):
    if n <= 1:
        return False
    elif n == 2:
        return True
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True
    
def solution(n, k):
    answer = 0
    s = ""

    while n > 0:
        s = str(n % k) + s
        n //= k

    i = 0
    while i < len(s):
        if s[i] == '0':
            i += 1
            continue

        j = i + 1
        while j < len(s) and s[j] != '0':
            j += 1

        number = int(s[i:j])
        if is_prime(number):
            answer += 1

        i = j

    return answer